src/components/main-content.tsx

```diff
@@ -3,14 +3,16 @@ import * as React from "react";
 import { BaseComponent, IBaseProps } from "./base";
 
 import "./main-content.sass";
-import { InvestigateDialogComponent } from "./investigate-dialog";
 import { PopulationsSpaceComponent } from "./spaces/populations-space";
 
 interface IProps extends IBaseProps {}
 interface IState {}
 
 const SpaceComponents: any = {
```

Comment:
This should probably be typed with something like:

```
type SpaceTypeToComponent = {
  [key in Space]: SpaceComponent | undefined
}
```

Which would presume we have some kind of union'd type for all SpaceComponents (just `PopulationsSpaceComponent` at the moment).

Also, I prefer to capitalize constants as SPACE_COMPONENTS, though we haven't been consistent on this.

@sfentress, do you have any thoughts on this?

Reply:
Ah, how does that work?  So we're specifying that the key has to be of type Space and the key's value has to be of type SpaceComponent (or undefined)?  How do we map the keys to the desired component?

Reply:
Or is the idea that we then make an object of type SpaceTypeToComponent and do the mapping there with  typescript enforcing the nice and neat key/value type rules for us?

Reply:
Yes, the second one! So once we had the type we could do:

```
const spaceTypeToComponent: SpaceTypeToComponent = {
  populations: PopulationsSpaceComponent
  dna: PopulationsSpaceComponent
  ...
  none: undefined
}
```

Also, to make the type def above that I worked we would use something like:

```
type SpaceComponent = PopulationSpaceComponent; // Eventually union all space components
```

`[key in Space]` is called a mapped type, and has the added benefit of making sure that *every* element in space is mapped to a component. So if you add a new `Space`, you'll get a compiler error to remind you to expand the map. You can write `[key in Space]?` instead if you don't want all types mapped - this just means that you'll need to check for undefined outputs from the map.

Happy to discuss more tomorrow if you'd like!

Reply:
Anyway, I see the idea.  I'm getting one, probably simple, error.  I DMed you with more info.

Reply:
On the topic of formatting constants, it's obviously an area over which strong opinions have been voiced over the years. Personally, I find the `ALL_CAPS_CONSTANT` formatting approach makes it seem like the constants are yelling at and potentially overshadowing the surrounding code. I would lean to a more soft-spoken convention like `kCamelCaseConstant`, but I'm fine with either.

Reply:
You know, especially now that most of our variables are constants anyway, all caps might be overkill. Is it possible that we don't need either?

Anyway, I'm fine sticking with `kCamelCase` for now. But please read all my code as yells to compensate.

Reply:
Also prefer the camelCase constant - the all caps is a little intrusive when constants start piling up (which in ES6 is everywhere).

Reply:
I would reserve the constant formatting for true constants. ES6 constants, e.g. `const { foo } = bar;` are `const` in the sense that they don't change during the life of a single function, but they're expected to have different values at different times in the running of the application.